File: planet-examples/src/main/java/org/saar/minecraft/chunk/ChunkLights.java
package org.saar.minecraft.chunk;

import org.saar.minecraft.Block;
import org.saar.minecraft.Chunk;
import org.saar.minecraft.EmptyChunk;
import org.saar.minecraft.LightPropagation;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

public class ChunkLights {

    private final Chunk chunk;

    private final byte[] lights = new byte[16 * 16 * 256];

    public ChunkLights(Chunk chunk) {
        this.chunk = chunk;
        Arrays.fill(this.lights, (byte) 0xFF);
    }

    private static int index(int x, int y, int z) {
        return ((x & 0xF) << 12) | ((z & 0xF) << 8) | (y & 0xFF);
    }

    private int calculateLight(int x, int y, int z) {
        final Block block = this.chunk.getBlock(x, y, z);
        final LightPropagation p = block.getLightPropagation();
        if (block.isTransparent()) {
            int max = 0;
            max = Math.max(max, this.chunk.getLight(x, y + 1, z) - p.getDown());
            max = Math.max(max, this.chunk.getLight(x, y - 1, z) - p.getSide());
            max = Math.max(max, this.chunk.getLight(x - 1, y, z) - p.getSide());
            max = Math.max(max, this.chunk.getLight(x + 1, y, z) - p.getSide());
            max = Math.max(max, this.chunk.getLight(x, y, z - 1) - p.getSide());
            max = Math.max(max, this.chunk.getLight(x, y, z + 1) - p.getSide());
            return max;
        }
        return 0;
    }

    public void recalculateLight() {
        Arrays.fill(this.lights, (byte) 0);

        final Queue<LightNode> spreadBfs = new LinkedList<>();
        for (int x = 0; x < 16; x++) {
            for (int z = 0; z < 16; z++) {
                if (this.chunk.getBlock(x, 0xFF, z).isTransparent()) {
                    spreadBfs.add(new LightNode(x, 0xFF, z, 0xFF));
                }
            }
        }

        spreadLight(spreadBfs);
    }

    public void updateLight(int x, int y, int z) {
        final int light = calculateLight(x, y, z);
        updateLight(x, y, z, light);
    }

    private void updateLight(int x, int y, int z, int light) {
        final Queue<LightNode> spreadBfs = new LinkedList<>();
        final Queue<LightNode> removeBfs = new LinkedList<>();

        if (this.chunk.getBlock(x, y, z).isTransparent()) {
            spreadBfs.add(new LightNode(x, y, z, light));
            this.chunk.setLight(x, y, z, (byte) light);
        } else {
            removeBfs.add(new LightNode(x, y, z, getLight(x, y, z)));
            this.chunk.setLight(x, y, z, (byte) 0);
        }

        removeLight(spreadBfs, removeBfs);
        spreadLight(spreadBfs);
    }


    private void removeLight(Queue<LightNode> addBfs, Queue<LightNode> remBfs) {
        while (!remBfs.isEmpty()) {
            final LightNode n = remBfs.poll();
            removeLightDown(addBfs, remBfs, n.x, n.y - 1, n.z, n);
            removeLightSide(addBfs, remBfs, n.x, n.y + 1, n.z, n);
            removeLightSide(addBfs, remBfs, n.x + 1, n.y, n.z, n);
            removeLightSide(addBfs, remBfs, n.x - 1, n.y, n.z, n);
            removeLightSide(addBfs, remBfs, n.x, n.y, n.z + 1, n);
            removeLightSide(addBfs, remBfs, n.x, n.y, n.z - 1, n);
        }
    }

    private void removeLightDown(Queue<LightNode> addBfs, Queue<LightNode> remBfs, int x, int y, int z, LightNode n) {
        final Block block = this.chunk.getBlock(x, y, z);
        final LightPropagation p = block.getLightPropagation();
        final int spread = Math.max(n.level - p.getDown(), 0);
        removeLight(addBfs, remBfs, x, y, z, spread);
    }

    private void removeLightSide(Queue<LightNode> addBfs, Queue<LightNode> remBfs, int x, int y, int z, LightNode n) {
        final Block block = this.chunk.getBlock(x, y, z);
        final LightPropagation p = block.getLightPropagation();
        final int spread = Math.max(n.level - p.getSide(), 0);
        removeLight(addBfs, remBfs, x, y, z, spread);
    }

    private void spreadLight(Queue<LightNode> spreadBfs) {
        while (!spreadBfs.isEmpty()) {
            final LightNode n = spreadBfs.poll();
            spreadLightDown(spreadBfs, n.x, n.y - 1, n.z, n);
            spreadLightSide(spreadBfs, n.x, n.y + 1, n.z, n);
            spreadLightSide(spreadBfs, n.x + 1, n.y, n.z, n);
            spreadLightSide(spreadBfs, n.x - 1, n.y, n.z, n);
            spreadLightSide(spreadBfs, n.x, n.y, n.z + 1, n);
            spreadLightSide(spreadBfs, n.x, n.y, n.z - 1, n);
        }
    }

    private void spreadLightDown(Queue<LightNode> addBfs, int x, int y, int z, LightNode n) {
        final Block block = this.chunk.getBlock(x, y, z);
        final LightPropagation p = block.getLightPropagation();
        final int spread = Math.max(n.level - p.getDown(), 0);
        spreadLight(addBfs, x, y, z, spread);
    }

    private void spreadLightSide(Queue<LightNode> addBfs, int x, int y, int z, LightNode n) {
        final Block block = this.chunk.getBlock(x, y, z);
        final LightPropagation p = block.getLightPropagation();
        final int spread = Math.max(n.level - p.getSide(), 0);
        spreadLight(addBfs, x, y, z, spread);
    }

    private void spreadLight(Queue<LightNode> addBfs, int x, int y, int z, int spread) {
        if (y < 0 || y > 255 || this.chunk.getRelativeChunk(x, z) == EmptyChunk.getInstance()) return;
        if (spread < 0) return;

        final int light = this.chunk.getLight(x, y, z);
        final Block block = this.chunk.getBlock(x, y, z);
        if (block.isTransparent() && light < spread) {
            this.chunk.setLight(x, y, z, (byte) spread);

            addBfs.add(new LightNode(x, y, z, spread));
        }
    }

    private void removeLight(Queue<LightNode> addBfs, Queue<LightNode> remBfs, int x, int y, int z, int spread) {
        if (y < 0 || y > 255 || this.chunk.getRelativeChunk(x, z) == EmptyChunk.getInstance()) return;

        final int light = this.chunk.getLight(x, y, z);
        if (light != 0 && light <= spread) {
            this.chunk.setLight(x, y, z, (byte) 0);

            remBfs.add(new LightNode(x, y, z, spread));
        } else if (light > spread) {
            addBfs.add(new LightNode(x, y, z, light));
        }
    }

    public int getLight(int x, int y, int z) {
        final int index = index(x, y, z);
        return getLight(index);
    }

    private int getLight(int index) {
        return this.lights[index] & 0xFF;
    }

    public void setLight(int x, int y, int z, byte level) {
        final int index = index(x, y, z);
        this.lights[index] = level;
    }

    private static class LightNode {
        public final int x;
        public final int y;
        public final int z;
        public final int level;

        public LightNode(int x, int y, int z, int level) {
            this.x = x;
            this.y = y;
            this.z = z;
            this.level = level;
        }

        @Override
        public String toString() {
            return "LightNode{" +
                    "x=" + x +
                    ", y=" + y +
                    ", z=" + z +
                    ", level=" + level +
                    '}';
        }
    }
}

File: planet-examples/src/main/java/org/saar/minecraft/chunk/ChunkLights.kt
package org.saar.minecraft.chunk

import org.saar.minecraft.Chunk
import org.saar.minecraft.EmptyChunk
import java.util.*
import kotlin.math.max

class ChunkLights(private val chunk: Chunk) {

    private val lights = ByteArray(16 * 16 * 256) { 0xFF.toByte() }

    private fun calculateLight(x: Int, y: Int, z: Int): Int {
        val block = this.chunk.getBlock(x, y, z)
        val p = block.lightPropagation
        if (block.isTransparent) {
            var max = 0
            max = max(max, this.chunk.getLight(x, y + 1, z) - p.down)
            max = max(max, this.chunk.getLight(x, y - 1, z) - p.side)
            max = max(max, this.chunk.getLight(x - 1, y, z) - p.side)
            max = max(max, this.chunk.getLight(x + 1, y, z) - p.side)
            max = max(max, this.chunk.getLight(x, y, z - 1) - p.side)
            max = max(max, this.chunk.getLight(x, y, z + 1) - p.side)
            return max
        }
        return 0
    }

    fun recalculateLight() {
        this.lights.fill(0.toByte())

        val spreadBfs = LinkedList<LightNode>()
        for (x in 0..15) {
            for (z in 0..15) {
                if (this.chunk.getBlock(x, 0xFF, z).isTransparent) {
                    spreadBfs.add(LightNode(x, 0xFF, z, 0xFF))
                }
            }
        }

        spreadLight(spreadBfs)
    }

    fun updateLight(x: Int, y: Int, z: Int) {
        val light = calculateLight(x, y, z)
        updateLight(x, y, z, light)
    }

    private fun updateLight(x: Int, y: Int, z: Int, light: Int) {
        val spreadBfs = LinkedList<LightNode>()
        val removeBfs = LinkedList<LightNode>()

        if (this.chunk.getBlock(x, y, z).isTransparent) {
            spreadBfs.add(LightNode(x, y, z, light))
            this.chunk.setLight(x, y, z, light.toByte())
        } else {
            removeBfs.add(LightNode(x, y, z, getLight(x, y, z)))
            this.chunk.setLight(x, y, z, 0.toByte())
        }

        removeLight(spreadBfs, removeBfs)
        spreadLight(spreadBfs)
    }


    private fun removeLight(addBfs: Queue<LightNode>, remBfs: Queue<LightNode>) {
        while (!remBfs.isEmpty()) {
            val n = remBfs.poll()
            removeLightDown(addBfs, remBfs, n.x, n.y - 1, n.z, n)
            removeLightSide(addBfs, remBfs, n.x, n.y + 1, n.z, n)
            removeLightSide(addBfs, remBfs, n.x + 1, n.y, n.z, n)
            removeLightSide(addBfs, remBfs, n.x - 1, n.y, n.z, n)
            removeLightSide(addBfs, remBfs, n.x, n.y, n.z + 1, n)
            removeLightSide(addBfs, remBfs, n.x, n.y, n.z - 1, n)
        }
    }

    private fun removeLightDown(
        addBfs: Queue<LightNode>,
        remBfs: Queue<LightNode>,
        x: Int,
        y: Int,
        z: Int,
        n: LightNode
    ) {
        val block = this.chunk.getBlock(x, y, z)
        val p = block.lightPropagation
        val spread = max(n.level - p.down, 0)
        removeLight(addBfs, remBfs, x, y, z, spread)
    }

    private fun removeLightSide(
        addBfs: Queue<LightNode>,
        remBfs: Queue<LightNode>,
        x: Int,
        y: Int,
        z: Int,
        n: LightNode
    ) {
        val block = this.chunk.getBlock(x, y, z)
        val p = block.lightPropagation
        val spread = max(n.level - p.side, 0)
        removeLight(addBfs, remBfs, x, y, z, spread)
    }

    private fun spreadLight(spreadBfs: Queue<LightNode>) {
        while (!spreadBfs.isEmpty()) {
            val n = spreadBfs.poll()
            spreadLightDown(spreadBfs, n.x, n.y - 1, n.z, n)
            spreadLightSide(spreadBfs, n.x, n.y + 1, n.z, n)
            spreadLightSide(spreadBfs, n.x + 1, n.y, n.z, n)
            spreadLightSide(spreadBfs, n.x - 1, n.y, n.z, n)
            spreadLightSide(spreadBfs, n.x, n.y, n.z + 1, n)
            spreadLightSide(spreadBfs, n.x, n.y, n.z - 1, n)
        }
    }

    private fun spreadLightDown(addBfs: Queue<LightNode>, x: Int, y: Int, z: Int, n: LightNode) {
        val block = this.chunk.getBlock(x, y, z)
        val p = block.lightPropagation
        val spread = max(n.level - p.down, 0)
        spreadLight(addBfs, x, y, z, spread)
    }

    private fun spreadLightSide(addBfs: Queue<LightNode>, x: Int, y: Int, z: Int, n: LightNode) {
        val block = this.chunk.getBlock(x, y, z)
        val p = block.lightPropagation
        val spread = max(n.level - p.side, 0)
        spreadLight(addBfs, x, y, z, spread)
    }

    private fun spreadLight(addBfs: Queue<LightNode>, x: Int, y: Int, z: Int, spread: Int) {
        if (y !in 0..255 || this.chunk.getRelativeChunk(x, z) === EmptyChunk) return
        if (spread < 0) return

        val light = this.chunk.getLight(x, y, z)
        val block = this.chunk.getBlock(x, y, z)
        if (block.isTransparent && light < spread) {
            this.chunk.setLight(x, y, z, spread.toByte())

            addBfs.add(LightNode(x, y, z, spread))
        }
    }

    private fun removeLight(addBfs: Queue<LightNode>, remBfs: Queue<LightNode>, x: Int, y: Int, z: Int, spread: Int) {
        if (y !in 0..255 || this.chunk.getRelativeChunk(x, z) === EmptyChunk) return

        val light = this.chunk.getLight(x, y, z)
        if (light != 0 && light <= spread) {
            this.chunk.setLight(x, y, z, 0.toByte())

            remBfs.add(LightNode(x, y, z, spread))
        } else if (light > spread) {
            addBfs.add(LightNode(x, y, z, light))
        }
    }

    fun getLight(x: Int, y: Int, z: Int): Int {
        val index: Int = index(x, y, z)
        return getLight(index)
    }

    private fun getLight(index: Int): Int {
        return this.lights[index].toInt() and 0xFF
    }

    fun setLight(x: Int, y: Int, z: Int, level: Byte) {
        val index: Int = index(x, y, z)
        this.lights[index] = level
    }

    private class LightNode(val x: Int, val y: Int, val z: Int, val level: Int) {
        override fun toString(): String {
            return "LightNode{" +
                    "x=" + x +
                    ", y=" + y +
                    ", z=" + z +
                    ", level=" + level +
                    '}'
        }
    }

    companion object {
        private fun index(x: Int, y: Int, z: Int): Int {
            return ((x and 0xF) shl 12) or ((z and 0xF) shl 8) or (y and 0xFF)
        }
    }
}
